python-package/xgboost/core.py

```diff
@@ -365,6 +423,33 @@ def _init_from_npy2d(self, mat, missing, nthread):
                 ctypes.byref(self.handle),
                 nthread))
 
+    def _init_from_dt(self, data, nthread):
+        """
+        Initialize data from a DataTable
+        """
+        cols = []
+        ptrs = (ctypes.c_void_p * data.ncols)()
+        for icol in range(data.ncols):
+            col = data.internal.column(icol)
+            cols.append(col)
+            # int64_t (void*)
+            ptr = col.data_pointer
+            ptrs[icol] = ctypes.c_void_p(ptr)
+
+        # always return stypes for dt ingestion
+        feature_stypes = (ctypes.c_wchar_p * data.ncols)()
+        for icol in range(data.ncols):
+            feature_stypes[icol] = ctypes.c_wchar_p(data.stypes[icol].name)
```

Comment:
Can we combine this loop with the loop above, since both loops loop from 0 to `data.ncols`.

Reply:
Also, let us re-consider the choice of `c_wchar_p` (see my comment in `c_api.h`). I think we can use `c_char_p` here by using `string.encode` to obtain UTF-8 strings.